cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/batch/preview/LimitingInputSplit.java

```diff
@@ -0,0 +1,141 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.etl.batch.preview;
+
+import org.apache.hadoop.conf.Configurable;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.io.Writable;
+import org.apache.hadoop.io.serializer.Deserializer;
+import org.apache.hadoop.io.serializer.SerializationFactory;
+import org.apache.hadoop.io.serializer.Serializer;
+import org.apache.hadoop.mapreduce.InputSplit;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * An {@link InputSplit} that delegates to another {@link InputSplit} and also carries record limit information.
+ */
+public class LimitingInputSplit extends InputSplit implements Writable, Configurable {
+
+  private List<InputSplit> inputSplits;
+  private int recordLimit;
+  private Configuration conf;
+  private long length;
+  private String[] locations;
+
+  @SuppressWarnings("unused")
+  public LimitingInputSplit() {
+    // no-op, for deserialization
+  }
+
+  LimitingInputSplit(Configuration conf, List<InputSplit> inputSplits, int recordLimit) throws IOException {
+    this.conf = conf;
+    this.inputSplits = inputSplits;
+    this.recordLimit = recordLimit;
+    initialize();
+  }
+
+  private void initialize() throws IOException {
+    try {
+      long length = 0L;
+      Set<String> locations = new LinkedHashSet<>();
+      for (InputSplit split : inputSplits) {
+        length += split.getLength();
+        String[] splitLocations = split.getLocations();
+        if (splitLocations != null) {
+          locations.addAll(Arrays.asList(splitLocations));
+        }
+      }
+      this.length = length;
+      this.locations = locations.toArray(new String[0]);
+    } catch (InterruptedException e) {
+      throw new IOException("Interrupted during initialization", e);
+    }
+  }
+
+  List<InputSplit> getInputSplits() {
+    return inputSplits;
+  }
+
+  int getRecordLimit() {
+    return recordLimit;
+  }
+
+  @Override
+  public long getLength() {
+    return length;
+  }
+
+  @Override
+  public String[] getLocations() {
+    return locations;
+  }
+
+  @Override
+  public void write(DataOutput out) throws IOException {
+    out.writeInt(recordLimit);
+    out.writeInt(inputSplits.size());
+    for (InputSplit split : inputSplits) {
+      Text.writeString(out, split.getClass().getName());
```

Comment:
since InputSplit is writable, can this just delegate to the split? Or do we need the classname and to use Serializer for some other reason?

Reply:
nevermind, was looking at mapred.InputSplit. Don't know why they didn't make it part of the InputSplit contract, seems like it would always be needed...